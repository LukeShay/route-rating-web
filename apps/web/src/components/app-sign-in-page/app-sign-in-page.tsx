import { Component, Host, h } from '@stencil/core'

import { HTMLEvent } from '@routerating/components'
import authStore from '../../stores/auth'
import { login } from '@routerating/requests'
import toastStore from '../../stores/toast'

interface Refs {
  email?: HTMLRrInputElement
  password?: HTMLRrInputElement
}

@Component({
  tag: 'app-sign-in-page',
  styleUrl: 'app-sign-in-page.scss',
  shadow: true,
})
export class AppSignInPage {
  private refs: Refs = {}

  render() {
    return (
      <Host>
        <app-form>
          <h1 class="centered">Sign In</h1>
          <form onSubmit={this.handleSubmit}>
            <rr-input
              label="Email"
              type="text"
              required
              ref={(element?: HTMLRrInputElement): void => {
                this.refs.email = element
              }}
            />
            <rr-input
              label="Password"
              type="password"
              required
              ref={(element?: HTMLRrInputElement): void => {
                this.refs.password = element
              }}
            />
            <div class="centered">
              <button type="submit" class="purple-button">
                Sign In
              </button>
            </div>
          </form>
        </app-form>
        <div class="centered">
          <stencil-route-link url="/signup" exact={true}>
            Create an account
          </stencil-route-link>
        </div>
      </Host>
    )
  }

  private handleSubmit = async (event: HTMLEvent): Promise<void> => {
    event.preventDefault()

    const response = await login(
      'http://localhost:3000/dev',
      (await this.refs.email?.getValue()) ?? '',
      (await this.refs.password?.getValue()) ?? ''
    )

    if (response.ok) {
      authStore.authToken = response.headers.get('Authorization')
      authStore.refreshToken = response.headers.get('Refresh')
      authStore.signedIn = response.headers.get('Authorization') !== null

      toastStore.message = 'You have been signed in.'
    } else {
      toastStore.message =
        'Unable to sign in. Verify your sign in information is correct.'
    }
  }
}
