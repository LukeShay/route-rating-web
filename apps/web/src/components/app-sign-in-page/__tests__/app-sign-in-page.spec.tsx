import { AppSignInPage } from '../app-sign-in-page'
import { newSpecPage } from '@stencil/core/testing'

describe('app-sign-in-page', () => {
  it('should match snapshot', async () => {
    const page = await newSpecPage({
      components: [AppSignInPage],
      html: `<app-sign-in-page></app-sign-in-page>`,
    })
    expect(page.root).toMatchSnapshot()
  })
})
