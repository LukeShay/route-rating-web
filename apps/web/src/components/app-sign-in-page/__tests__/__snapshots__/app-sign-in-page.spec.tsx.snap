// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`app-sign-in-page should match snapshot 1`] = `
<app-sign-in-page>
  <mock:shadow-root>
    <app-form>
      <h1 class="centered">
        Sign In
      </h1>
      <form>
        <rr-input label="Email" required="" type="text"></rr-input>
        <rr-input label="Password" required="" type="password"></rr-input>
        <button class="purple-button">
          Sign In
        </button>
      </form>
    </app-form>
  </mock:shadow-root>
</app-sign-in-page>
`;
